Fixes #844 #5725 - correct hosts count in puppetclass

tbrisker opened the pull request.

tbrisker committed:
Fixes #844 #5725 - correct hosts count in puppetclass

tbrisker commented:
rebased and added migration

tbrisker commented:
[test]

isratrade commented on lib/tasks/fix_cached_counters.rake:
```diff
@@ -5,7 +5,7 @@ task :fix_cached_counters => :environment do
     puts "#{cl} corrected"
   end
-  Puppetclass.all.each{|el| Puppetclass.reset_counters(el.id, :hosts, :hostgroups, :lookup_keys)}
+  Puppetclass.all.each{|el| Puppetclass.reset_counters(el.id, :hostgroups, :lookup_keys)}
```
@tbrisker, why was `:hosts` removed?

tbrisker replied:
because it is no longer a cached_counter

domcleal commented on app/models/puppetclass.rb:
```diff
@@ -46,6 +46,10 @@ def to_param
   end
 
+  def update_hosts_count(relation = nil)
+    update_attribute(:hosts_count, Host::Managed.search_for("class = #{name}").count)
```
unscoped hosts?  Default scope might be in play.

ditto in a few other places I think, this seems to be pretty unaware of default scopes (yeah, yeah, #6878).

tbrisker replied:
@domcleal - Seems like using unscoped.search_for still won't get te expected results if there are any taxonomies in effect because of [how `search_by_puppetclass` works](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/hostext/search.rb#L100). 
Looks like I will need to look deeper into this issue.

dLobatog commented:
Thanks @tbrisker for the patch. I've set this to Waiting on contributor since it needs rebasing and addressing the issue with `default_scope` @domcleal mentioned.

dLobatog commented:
@tbrisker Sorry :bow: but this needs a rebase against theforeman/foreman:develop

dLobatog commented on app/models/host/managed.rb:
```diff
@@ -919,4 +922,23 @@ def update_lookup_value_fqdn_matchers
+  end
+
+  def update_all_puppetclasses_hosts_count
+    unless puppetclasses.nil?
```
`if puppetclasses.present?` reads a bit better

dLobatog commented on app/models/host/managed.rb:
```diff
@@ -919,4 +922,23 @@ def update_lookup_value_fqdn_matchers
+      puppetclasses.each { |p| p.update_hosts_count }
+    end
+    hostgroup.update_all_puppetclasses_hosts_count unless hostgroup.nil?
+    unless config_groups.nil?
```
Same as above

dLobatog commented on app/models/puppetclass.rb:
```diff
@@ -82,6 +82,23 @@ def klass
+    ids = host_ids
+    ids += all_hostgroups.flat_map(&:host_ids)
+    ids += config_groups.joins(:host_config_groups).where("host_config_groups.host_type='Host::Base'").pluck('host_config_groups.host_id') unless config_group_classes.empty?
+    Host::Managed.unscoped.where(:id => ids).uniq
```
Maybe it's better to `uniq` the ids before searching rather than after so we only search for the needed hosts?

dLobatog commented on app/models/puppetclass.rb:
```diff
@@ -82,6 +82,23 @@ def klass
 
+  def all_hostgroups
+    ids = hostgroup_ids
+    ids += config_groups.joins(:host_config_groups).where("host_config_groups.host_type='Hostgroup'").pluck('host_config_groups.host_id') unless config_group_classes.empty?
```
Make a separate method with an argument `host_type` and a clear title so you can call this query from all_hosts and all_hostgroups. I know what it does but it takes a few seconds to read.

dLobatog commented on app/models/host/managed.rb:
```diff
@@ -919,4 +922,23 @@ def update_lookup_value_fqdn_matchers
     LookupValue.where(:match => "fqdn=#{fqdn_was}").update_all(:match => lookup_value_match)
   end
 
+  def update_puppetclass_hosts_count(puppetclass)
```
This method appears verbatim in `config_group` and `hostgroup`

dLobatog commented:
@tbrisker Thanks, it works well but the `update_` methods are very similar and repetitive. Aside from the comments in the code, could you refactor the `update_` stuff out into a concern shared by these 3 classes? You can get the model name dynamically to change behavior depending if it's a hostgroup/host/config_group.

dLobatog commented on lib/tasks/puppet.rake:
```diff
@@ -217,4 +217,14 @@ namespace :puppet do
+  task :fix_hosts_counts => :environment do
+    if Puppetclass.count > 0
+      User.current = User.anonymous_admin
+      Puppetclass.all.each do |pc|
```
`Puppetclass.all.each(&:update_hosts_count)` is faster,  https://speakerdeck.com/sferik/writing-fast-ruby?slide=15

dLobatog commented on test/unit/puppetclass_test.rb:
```diff
@@ -76,4 +76,36 @@ class PuppetclassTest < ActiveSupport::TestCase
+    pc = FactoryGirl.create(:puppetclass)
+    hg = FactoryGirl.create(:hostgroup, :with_config_group)
+    cg = hg.config_groups.first
+    cg.puppetclasses << pc
```
Please use meaningful variables, pc/hg/cg are not instantly obvious for everyone.

dLobatog commented on app/models/puppetclass.rb:
```diff
@@ -82,6 +82,32 @@ def klass
 
+  # return host ids from config groups by type
+  def host_ids_from_config_groups(host_type)
+    cg = config_groups.joins(:host_config_groups)
```
Why `cg` , especially since these are `host_ids`?

dLobatog commented:
I think `update_puppetclass_hosts_count`, and `update_all_puppetclasses_hosts_count` in HostCommon, HostManaged, ConfigGroup, and Hostgroup are irrelevant, because adding/removing a puppetclass to any of these objects is going to trigger `update_hosts_count` on the Puppetclass model. So if I'm not wrong you can get rid of all these methods.

The exception is adding/removing a Hostgroup or a ConfigGroup from a Host. The Puppetclass model doesn't know about that yet it needs to be updated. For that you can keep a method:

``` ruby
# Update counters for all Puppetclasses affected indirectly
  def update_all_puppetclasses_hosts_count
    puppetclasses.each(&:update_hosts_count)
  end
```

and run it on `has_many_hosts :after_add, :after_remove` on ConfigGroup and Hostgroup.

The reasoning is that these two classes are the only ones that can change the host count for a Puppetclass **without** the Puppetclass knowing about it.

I've sent you a PR (untested) - https://github.com/tbrisker/foreman/pull/1 - with the proposed changes to this branch. This logic is a bit confusing and not central to Foreman core business in my opinion, for Hostgroups shouldn't gem ancestry take care of this? Maybe it's a good time to rethink how to take all this kind of logic out of the main app.

Could you add some comments on how you've tested this thoroughly too? It was a bit time consuming to figure out what this was doing and how to test it.

Thanks @tbrisker !

tbrisker commented:
@eLobato I added many tests for (hopefully) all possible puppetclass inheritance routes. 
Also cleaned up the code some more and addressed the issues.
Please re-review.

dLobatog commented on app/models/puppetclass.rb:
```diff
@@ -10,12 +10,12 @@ class Puppetclass < ActiveRecord::Base
+  has_many :hostgroup_classes
+  has_many :hostgroups, :through => :hostgroup_classes, :dependent => :destroy
+  has_many :host_classes
+  has_many_hosts :through => :host_classes, :dependent => :destroy
```
Why have you moved the `:dependent => :destroy` away from the join-table classes (host_classes, etc) everywhere?

tbrisker replied:
`:dependent => :destroy` on a has-many through relationship causes the entry in the join table to be destroyed when deleting the relation, and fires the `:after_destroy` callback which doesn't fire by default (which is `:delete_all` - no callbacks fired). [See here](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#module-ActiveRecord::Associations::ClassMethods-label-Deleting+from+associations) for more details.

dLobatog commented:
@tbrisker Looks ready to merge to me, except for the dependent destroy part, which I don't understand, thanks! It's much better than the initial version in my opinion.

dLobatog commented on app/models/puppetclass.rb:
```diff
@@ -82,6 +82,33 @@ def klass
+  end
+
+  def update_total_hosts(relation = nil)
+    # binding.pry
```
No need for the binding.pry comment and `(relation = nil)`

dLobatog commented on lib/tasks/puppet.rake:
```diff
@@ -217,4 +217,12 @@ namespace :puppet do
+  task :fix_hosts_counts => :environment do
+    if Puppetclass.count > 0
+      User.current = User.anonymous_admin
+      Puppetclass.all.each(&:update_hosts_count)
```
`update_hosts_count` should be `update_total_hosts`

dLobatog commented on db/migrate/20140922084848_fix_puppetclass_hosts_counters.rb:
```diff
@@ -0,0 +1,8 @@
+class FixPuppetclassHostsCounters < ActiveRecord::Migration
```
The timestamp of this migration `20140922084848` is < than the timestamp of the other migration `20141109131448`. This means the user won't be able to run db:migrate because it'll try to trigger the `fix_hosts_counts` task without having the `total_hosts` column. Change this timestamp to something like `20141109131450` and that'd be fixed. (that's what I did to test the PR)

dLobatog commented:
@tbrisker Tested and it works well, thanks very much. I think it's really a lot of complexity for a counter but I can't think of a better way of doing this off the top of my head, maybe some gem does this kind of stuff. Please change the migration timestamp and the task as commented inline and :+1: to merge

dLobatog commented:
Merged as 83683ed0aeb6ab38d518837698d90c50b4adf65b , thanks @tbrisker , I know this took a lot of effort for a counter :smile: !
